### PR TITLE
`linera-views`: avoid generating duplicate GraphQL type names

### PR DESCRIPTION
**Pull request by Twey**
## Motivation

When we added the ability to do more complex queries on our `linera-views` data structures, we introduced several intermediate data types in the GraphQL that need unique type names.  GraphQL type names are a lot more syntactically restrictive than Rust type names, so we introduced a very basic name mangling scheme that is sufficient to cover all the cases originally covered by the `GraphQlView` derive macro and preserves human-readability. 

However, recently, we've seen an influx of issues in which people want to use more complex types in their `async_graphql::SimpleObject` types that cause these GraphQL type names to collide, resulting in a build error.

## Proposal

Introduce a more unique naming scheme by appending a hash of the Rust type name to the generated GraphQL type name.  This mangling scheme [isn't perfect](https://github.com/linera-io/linera-protocol/issues/1858), but should suffice for now until we get time to design something more appropriate.

## Test Plan

CI.

## Release Plan

- Need to bump the major/minor version number in the next release of the crates. (This is a breaking change to our GraphQL APIs.)

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-service-graphql-client/gql/service_schema.graphql [hunk 1]
```diff
@@ -140,19 +140,19 @@ type ChainStateExtendedView {
...
 	"""
 	Mailboxes used to receive messages indexed by their origin.
 	"""
-	inboxes: ReentrantCollectionView_Origin_InboxStateView!
+	inboxes: _12linera_views02_3a3a25reentrant_collection_view02_3a3a23ReentrantCollectionView01_3c12linera_views02_3a3a6common02_3a3a16ContextFromStore01_3c14linera_storage02_3a3a19ChainRuntimeContext01_3c14linera_storage02_3a3a10db_storage02_3a3a9DbStorage01_3c12linera_views02_3a3a6memory02_3a3a11MemoryStore02_2c2014linera_storage02_3a3a10db_storage02_3a3a9WallClock04_3e3e2c2012linera_views02_3a3a6memory02_3a3a11MemoryStore03_3e2c2012linera_chain02_3a3a10data_types02_3a3a6Origin02_2c2012linera_chain02_3a3a5inbox02_3a3a14InboxStateView01_3c12linera_views02_3a3a6common02_3a3a16ContextFromStore01_3c14linera_storage02_3a3a19ChainRuntimeContext01_3c14linera_storage02_3a3a10db_storage02_3a3a9DbStorage01_3c12linera_views02_3a3a6memory02_3a3a11MemoryStore02_2c2014linera_storage02_3a3a10db_storage02_3a3a9WallClock04_3e3e2c2012linera_views02_3a3a6memory02_3a3a11MemoryStore03_3e3e3e!
```
Comment (ma2bd) [inline linera-service-graphql-client/gql/service_schema.graphql:151]: Mmm ok, this is too much. At this point, how about hashing the typename and use hexadecimal digits? `Type_acbd3412baaaaaaaaabbbb000`.

Comment (ma2bd) [reply linera-service-graphql-client/gql/service_schema.graphql:151]: Also possible: a mix of GraphQL type + typename hash: `ReentrantCollectionView_Origin_InboxStateView_Nonnull_12312bdcasd`

Comment (Twey) [reply linera-service-graphql-client/gql/service_schema.graphql]: I think that's much better :)  I've updated this PR.

File: linera-views/src/graphql.rs [hunk 2]
```diff
@@ -23,7 +33,13 @@ impl<K: async_graphql::OutputType, V: async_graphql::OutputType> async_graphql::
     for Entry<K, V>
 {
     fn type_name() -> Cow<'static, str> {
-        format!("Entry_{}_{}", K::type_name(), V::type_name()).into()
+        format!(
+            "Entry_{}_{}_{:08x}",
+            K::type_name(),
```
Comment (ma2bd) [inline linera-views/src/graphql.rs:38]: Do you mean to replace `[]` and `!` in another PR?

Comment (Twey) [reply linera-views/src/graphql.rs]: I didn't, I just forgot :sweat_smile: I've added a commit, thanks for the catch!

File: linera-views/src/graphql.rs [hunk 3]
```diff
@@ -10,6 +10,31 @@ use crate::{
...
 // TODO(#1326): fix the proliferation of constraints from this module
 
+// TODO(#1858): come up with a better name-mangling scheme
+/// Mangle a GraphQL type into something that can be interpolated into a GraphQL type name
+fn mangle(type_name: impl AsRef<str>) -> String {
+    let mut mangled = String::new();
+
+    for c in type_name.as_ref().chars() {
+        match c {
+            '!' | '[' => (),
```
Comment (ma2bd) [inline linera-views/src/graphql.rs:20]: I was thinking we would to map `!` to `_NonNull`.

Comment (ma2bd) [reply linera-views/src/graphql.rs:20]: ^^ @Twey

Comment (Twey) [reply linera-views/src/graphql.rs:20]: Oh sorry I missed this one!
I don't think it's important to distinguish nullable/non-nullable in the type name.  These names are only for human-readability (now) and I don't think it's likely that a user will confuse a nullable type for another type that is the same modulo nullability.

Comment (ma2bd) [reply linera-views/src/graphql.rs:20]: Thanks. Sounds good!

**Pull request merged**